benchmarks/dynamo/common.py

```diff
@@ -388,16 +406,26 @@ def maybe_profile(*args, **kwargs):
                 if should_randomize_input
                 else example_inputs
             )
+            # need call mark_step to perform the computation
+            # on randomize_input. Otherwise the first call using the
+            # inputs will incur high penalty then the next one.
+            maybe_mark_step(args)
```

Comment:
i think we should be doing the mark step on the input boundary of the dynamo-xla backend, in order to ensure our graph-tracing starts on fresh xlatensors not xlatensors that have already accumulated some trace.  I think you're also working on that change in parallel.  Can we just get that working first and then not need this?

Reply:
@wconstab hmm, I think we still need it.

E.g., if we skip this mark_step call and do it inside the optimized function, then that optimized function will incur overhead of computing the randomized inputs. The other call we compare with does not have this overhead. 

Then the comparison would be unfair.

Reply:
We add `mark_step` or sync call in a few other places for the baseline. Since we trace on xla, the baseline is also running on XLA, we need explicitly sync tensors for it.

Reply:
@wconstab  I added a sync when we enter the optimized function.

But I think we may not need a mandate sync when we leave the function. Leave it as it is have the advantage that we potentially accumulate a large graph. Let me know if there are other concerns.

Reply:
ahh.. when we exit the optimized function, isn't the last thing we do to invoke an API that executes the cached xla graph?  that API may be equivalent to mark_step in all the ways that matter.  all of the tensors we return from the optimized function should be 
 - XLATensors
 - backed by devicedata, not IR value
 
 so any operations performed in eager land after one of our optimized functions returns should be starting fresh graphs.  If you returned XLAtensor backed by IR value from the optimized function, i don't think the xla runtime would behave the way you expect if you trace additional IR onto the graph and then execute it.

at least for the first design, i think it is best to break the XLA graph on entry/exit of the dynamo function

Reply:
yep, for now the output of the optimized function will most likely be device data. Adding a mark_step should not matter much since it's like a nop. I can add one to be safe.

Reply:
ok. i don't actually want to have an extra one. it might incur some small pointless overhead. it's probably safe to have it.  it's worth reviewing this part of the design with @JackCaoG and make sure we're not missing anything. but if we can prove the tensors we're returning from the execute-cached-graph api are fresh in all cases, i think we don't need the output mark step.

Reply:
This extra `mark_step` to sync the inputs is needed only for benchmarking. If we don't want it, we need make sure both runs (baseline and optimized) incur the same overhead of executing the graph that computes the inputs.

From other scenario (e.g. prod case if there is any in future), people don't need add such a call for sure.

Reply:
If we run the cached computation directly, no additional IR should be generated since we skip the tracing part.
If we run the model on lazy for the first time, we already call `xla_sync_multi` to sync all results of the models. The only IR that might not be materialized should be the IR seed.

I think we can use `torch_xla._XLAC. _xla_tensors_report(0, str(xla_device))` before the `mark_step` to confirm my theory. If there are additional pending graphs I felt like it might be an issue and we should address it.

Reply:
> This extra mark_step to sync the inputs is needed only for benchmarking. If we don't want it, we need make sure both runs (baseline and optimized) incur the same overhead of executing the graph that computes the inputs.

ok. for benchmarking reasons, you may be right. is mark_step synchronous though? i thought mark_step was just going to break the graph and schedule the execution.  you'd also need to ensure the execution finished.  also, i think there is a point in the dynamo benchmark where cuda.sync happens before/after measurements for this same reason. maybe that is where you should move this maybe_mark_step thing?

Reply:
`mark_step` is not synchronous by defualt, there a arg you can toggle the `wait` behavior. You are right that we actually need to wait for previous computation to finish. In lazy code we achieved that by trying to grab the deviceContext's lock. We are currently not doing that, so let me draft a quick pr to do something similar to https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1035 in our `_run_cached_graph` api.

Reply:
Instead of making `mark_step` synchronous here, I think it is better to make `_run_cached_graph` grab the device lock. This way we can overlap the dynamo bridges that prepare inputs to `_run_cached_graph` and the device execution.

Reply:
> also, i think there is a point in the dynamo benchmark where cuda.sync happens before/after measurements for this same reason.

yep, I think I know which of the code you mean. But one difference here is, for the cuda.sync, we want its execution time to be included to the benchmark execution time. But for this `mark_step`, we want to exclude it.

Reply:
now I think about it a bit more there are two cases

1. When we first time tracing the graph. This one is actually because we trace using the lazy and lock thing should already been taking care of.
2. When we run the cached graph, since we already run `torch_xla._XLAC._xla_sync_multi(args, [])`, we know that at least that input to the cached graph is materalized. However in a very rare chance that there are still some pending computations on XLA device. We can not execute more than 1 graph at any given time, so `_run_cahce_graph` should still try to grab the device lock before proceed to execution.

Reply:
https://github.com/pytorch/xla/pull/4195